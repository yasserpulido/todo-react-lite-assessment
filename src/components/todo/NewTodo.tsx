/** @jsxImportSource @emotion/react */
import React, { useState } from "react";
import { BaseSyntheticEvent } from "../../types/Events";
import { css } from "@emotion/react";

interface Props {
  createTodo: (a: string) => void;
}

const addInput = css({
  backgroundColor: "#fbc531",
  border: "2px solid #000",
  borderRadius: "4px",
  padding: "6px",
  marginRight: "8px",
  minWidth: "200px",
  outline: "none",
});

const addButton = css({
  backgroundColor: "#f5f6fa",
  border: "1px solid #95a5a6",
  cursor: "pointer",
  padding: "6px 16px",
  borderRadius: "4px",
  "&:hover": {
    backgroundColor: "#dcdde1",
  },
});

const NewTodo: React.FC<Props> = (props) => {
  const [text, setText] = useState("");

  const submitHandler = (event: BaseSyntheticEvent) => {
    event.preventDefault();

    if (text.trim().length === 0) {
      return;
    }

    props.createTodo(text);
    setText("");
  };

  const todoHandler = (event: BaseSyntheticEvent) => {
    setText(event.target.value);
  };

  return (
    <form onSubmit={submitHandler} aria-label="form">
      <input
        css={addInput}
        type="text"
        placeholder="Nueva tarea"
        value={text}
        onChange={todoHandler}
      />
      <button css={addButton} type="submit">
        Agregar
      </button>
    </form>
  );
};

export default NewTodo;
