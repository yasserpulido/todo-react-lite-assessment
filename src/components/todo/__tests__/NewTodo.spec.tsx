import { fireEvent, render, screen } from "@testing-library/react";
import NewTodo from "../NewTodo";

describe("NewTodo", () => {
  test("Should NewTodo form to match Snapshot", () => {
    // Act
    const { container } = render(<NewTodo createTodo={() => {}} />);

    // Assert
    expect(container).toMatchSnapshot();
  });

  test("Should EditTodo form render", () => {
    // Act
    render(<NewTodo createTodo={() => {}} />);
    const form = screen.getByRole("form");
    const input = screen.getByRole("textbox");
    const button = screen.getByRole("button");

    // Assert
    expect(form).toBeInTheDocument();
    expect(input).toBeInTheDocument();
    expect(input).toHaveAttribute("type", "text");
    expect(button.innerHTML).toBe("Agregar");
  });

  test("Should EditTodo form submit and call createTodo method", () => {
    // Arrange
    const text = "Hacer una prueba";
    const createTodo = jest.fn();

    // Act
    render(<NewTodo createTodo={createTodo} />);

    const form = screen.getByRole("form");
    const input = screen.getByRole("textbox");

    fireEvent.input(input, { target: { value: text } });
    fireEvent.submit(form);

    // Assert
    expect(createTodo).toHaveBeenCalled();
    expect(createTodo).toHaveBeenCalledTimes(1);
    expect(createTodo).toHaveBeenCalledWith(text);
  });

  test("Should not EditTodo form submit", () => {
    // Arrange
    const createTodo = jest.fn();

    // Act
    render(<NewTodo createTodo={createTodo} />);
    const form = screen.getByRole("form");
    fireEvent.submit(form);

    // Assert
    expect(createTodo).not.toHaveBeenCalled();
  });
});
