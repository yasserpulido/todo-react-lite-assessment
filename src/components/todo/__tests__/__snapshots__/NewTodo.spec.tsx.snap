// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewTodo Should NewTodo form to match Snapshot 1`] = `
<div>
  <form
    aria-label="form"
  >
    <input
      class="css-fmchn6-NewTodo"
      placeholder="Nueva tarea"
      type="text"
      value=""
    />
    <button
      class="css-1rpjkl-NewTodo"
      type="submit"
    >
      Agregar
    </button>
  </form>
</div>
`;
